File: lib/src/features/quiz/data/models/question.dart
import '../../domain/entities/question.dart' as entities;

/// Question data model for data layer
///
/// This model extends the domain entity with data-specific functionality
/// like JSON serialization and additional metadata for persistence.
class QuestionModel extends entities.Question {
  const QuestionModel({
    required super.id,
    required super.text,
    required super.type,
    required super.difficulty,
    required super.options,
    required super.correctAnswer,
    super.explanation,
    super.tags = const [],
    required super.subject,
    required super.course,
    required super.topic,
    super.subtopic,
    required super.knowledgeType,
    super.points = 10,
    super.initialConfidence = 0.5,
  });

  /// Create from domain entity
  factory QuestionModel.fromEntity(entities.Question question) {
    return QuestionModel(
      id: question.id,
      text: question.text,
      type: question.type,
      difficulty: question.difficulty,
      options: question.options,
      correctAnswer: question.correctAnswer,
      explanation: question.explanation,
      tags: question.tags,
      subject: question.subject,
      course: question.course,
      topic: question.topic,
      subtopic: question.subtopic,
      knowledgeType: question.knowledgeType,
      points: question.points,
      initialConfidence: question.initialConfidence,
    );
  }

  /// Convert to domain entity
  entities.Question toEntity() {
    return entities.Question(
      id: id,
      text: text,
      type: type,
      difficulty: difficulty,
      options: options,
      correctAnswer: correctAnswer,
      explanation: explanation,
      tags: tags,
      subject: subject,
      course: course,
      topic: topic,
      subtopic: subtopic,
      knowledgeType: knowledgeType,
      points: points,
      initialConfidence: initialConfidence,
    );
  }

  /// Create from JSON (for API integration)
  factory QuestionModel.fromJson(Map<String, dynamic> json) {
    final knowledgeType = json['knowledgeType']?.toString() ?? 'general';
    
    return QuestionModel(
      id: json['id'] as String,
      text: json['prompt']?.toString() ?? json['text'] as String,
      type: entities.QuestionType.fromJson(json['type']?.toString()),
      difficulty: json['difficulty'] != null 
        ? entities.DifficultyLevel.values.firstWhere(
            (e) => e.name == json['difficulty'],
            orElse: () => entities.DifficultyLevel.intermediate,
          )
        : entities.DifficultyLevel.fromKnowledgeType(knowledgeType),
      options: List<String>.from(json['options'] ?? []),
      correctAnswer: json['correctAnswer'] as String,
      explanation: json['explanation'] as String?,
      tags: List<String>.from(json['tags'] ?? json['metadata']?['tags'] ?? []),
      subject: json['subject']?.toString() ?? json['course']?.toString() ?? '',
      course: json['course']?.toString() ?? json['subject']?.toString() ?? '',
      topic: json['topic']?.toString() ?? 'general',
      subtopic: json['subtopic']?.toString(),
      knowledgeType: knowledgeType,
      points: json['points'] as int? ?? 10,
      initialConfidence: json['initialConfidence'] as double? ?? 0.5,
    );
  }

  /// Convert to JSON
  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'text': text,
      'type': type.name,
      'difficulty': difficulty.name,
      'options': options,
      'correctAnswer': correctAnswer,
      'explanation': explanation,
      'tags': tags,
      'subject': subject,
      'points': points,
      'initialConfidence': initialConfidence,
    };
  }
}

File: backend/app/models/question.py
from sqlalchemy import Column, Integer, String, Text, Float, DateTime, Boolean, JSON, ForeignKey
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from . import Base


class Question(Base):
    """Enhanced Question model with foreign keys for MAB system"""
    __tablename__ = "questions"

    # Primary fields
    id = Column(Integer, primary_key=True, autoincrement=True)
    question_id = Column(String(64), unique=True, index=True, nullable=False)

    # Foreign keys for categorization
    subtopic_id = Column(Integer, ForeignKey("subtopics.id", ondelete="SET NULL"), nullable=True, index=True)
    knowledge_type_id = Column(Integer, ForeignKey("knowledge_types.id", ondelete="SET NULL"), nullable=True, index=True)

    # Content
    text = Column(Text, nullable=False)  # Question text (prompt)
    type = Column(String(32), nullable=False, index=True)  # multiple_choice, true_false, fill_in_blank, matching

    # Options and answer
    options = Column(JSON, nullable=True)  # JSON array for multiple_choice: ["A", "B", "C", "D"]
    correct_answer = Column(String(500), nullable=False)  # Correct answer
    explanation = Column(Text, nullable=True)  # Explanation

    # For matching questions (future support)
    match_pairs = Column(JSON, nullable=True)  # JSON for matching: [{"left": "...", "right": "..."}, ...]

    # Legacy string fields (for backward compatibility during migration)
    course = Column(String(64), index=True, nullable=True)
    subject = Column(String(64), index=True, nullable=True)
    topic = Column(String(128), index=True, nullable=True)
    subtopic = Column(String(128), nullable=True)
    knowledge_type = Column(String(64), index=True, nullable=True)

    # Tags for additional categorization
    tags = Column(JSON, nullable=True)  # JSON array: ["NSAID", "pain", "aspirin"]

    # Difficulty
    difficulty = Column(String(32), index=True, default="intermediate")  # beginner, intermediate, advanced
    initial_confidence = Column(Float, default=0.5)  # For MAB algorithm

    # Scoring
    points = Column(Integer, default=10)

    # Metadata
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())
    is_active = Column(Boolean, default=True)

    # Relationships
    subtopic_rel = relationship("Subtopic", back_populates="questions")
    knowledge_type_rel = relationship("KnowledgeType", back_populates="questions")

    @property
    def options_json(self):
        """Backward compatibility property"""
        import json
        return json.dumps(self.options, ensure_ascii=False) if self.options else None

    def to_dict(self, include_relations=False):
        """Convert to dictionary for API responses"""
        result = {
            "id": self.question_id,
            "dbId": self.id,
            "prompt": self.text,
            "text": self.text,
            "type": self.type,
            "options": self.options,
            "correctAnswer": self.correct_answer,
            "explanation": self.explanation,
            "matchPairs": self.match_pairs,
            "subtopicId": self.subtopic_id,
            "knowledgeTypeId": self.knowledge_type_id,
            # Legacy fields
            "course": self.course,
            "subject": self.subject,
            "topic": self.topic,
            "subtopic": self.subtopic,
            "knowledgeType": self.knowledge_type,
            "tags": self.tags or [],
            "difficulty": self.difficulty,
            "initialConfidence": self.initial_confidence,
            "points": self.points,
            "isActive": self.is_active,
            "createdAt": self.created_at.isoformat() if self.created_at else None,
            "updatedAt": self.updated_at.isoformat() if self.updated_at else None,
        }

        if include_relations:
            if self.subtopic_rel:
                result["subtopicInfo"] = {
                    "id": self.subtopic_rel.id,
                    "name": self.subtopic_rel.name,
                    "displayName": self.subtopic_rel.display_name,
                }
                if self.subtopic_rel.topic:
                    result["topicInfo"] = {
                        "id": self.subtopic_rel.topic.id,
                        "name": self.subtopic_rel.topic.name,
                        "displayName": self.subtopic_rel.topic.display_name,
                    }
                    if self.subtopic_rel.topic.course:
                        result["courseInfo"] = {
                            "id": self.subtopic_rel.topic.course.id,
                            "name": self.subtopic_rel.topic.course.name,
                            "displayName": self.subtopic_rel.topic.course.display_name,
                        }
            if self.knowledge_type_rel:
                result["knowledgeTypeInfo"] = {
                    "id": self.knowledge_type_rel.id,
                    "name": self.knowledge_type_rel.name,
                    "displayName": self.knowledge_type_rel.display_name,
                }

        return result
